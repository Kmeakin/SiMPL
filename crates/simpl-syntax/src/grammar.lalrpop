use std::str::FromStr;
use crate::ast::{Symbol, Expr, LiteralExpr, VarExpr, LetExpr, Binding, LambdaExpr, AppExpr};

grammar;

// Literal expressions

pub LitExpr: LiteralExpr = {
    Bool => LiteralExpr::Bool(<>),
    Int => LiteralExpr::Int(<>),
    Float => LiteralExpr::Float(<>),
}

pub Bool: bool = {
    "true" => true,
    "false" => false,
}

pub Int: i64 = {
    <s:r"[0-9]+"> => i64::from_str(s).unwrap(),
}

pub Float: f64 = {
    <s:r"[0-9]+\.[0-9]+"> => f64::from_str(s).unwrap(),
}

// Variable references

pub Ident: Symbol = {
    <s:r"[a-zA-Z][a-zA-Z0-9_]*"> => <>.into(),
}

pub VarExpr: VarExpr = {
    Ident => VarExpr(<>),
}

// Let bindings

pub LetExpr: LetExpr = {
    "let" <bindings:Comma1<Binding>> "in" <body:Expr> ";"
        => LetExpr {bindings, body: box body},
}

pub Binding: Binding = {
    <var:Ident> "=" <val:Expr> => Binding {var, val: box val},
}

// Lambda abstraction

pub LambdaExpr: LambdaExpr = {
   "\\" "(" <args:Comma<Ident>> ")" "->" <body:Expr> ";"
        => LambdaExpr {args, body: box body},
}

// Function application

pub AppExpr: AppExpr = {
    <func:Expr> "(" <args:Comma<Expr>> ")" => AppExpr {func: box func, args},
}

// Expressions

pub Expr: Expr = {
    LitExpr => Expr::Lit(<>),
    VarExpr => Expr::Var(<>),
    LetExpr => Expr::Let(<>),
    LambdaExpr => Expr::Lambda(<>),
    AppExpr => Expr::App(<>),
}

#[inline]
Delim<T, D>: Vec<T> = {
    <v:(<T> <D>)*> <e:T?> => match e {
        None => v.iter().map(|x| x.0.clone()).collect(),
        Some(e) => {
            let mut v = v.iter().map(|x| x.0.clone()).collect::<Vec<_>>();
            v.push(e);
            v
        }
    }
}

#[inline]
Delim1<T, D>: Vec<T> = {
    <v:(<T> <D>)*> <e:T> => {
        let mut v = v.iter().map(|x| x.0.clone()).collect::<Vec<_>>();
        v.push(e);
        v
    }
}

#[inline]
Comma<T> = Delim<T, ",">;

#[inline]
Comma1<T> = Delim1<T, ",">;
