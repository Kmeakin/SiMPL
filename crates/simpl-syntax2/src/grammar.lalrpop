use std::str::FromStr;
use crate::ast::{Ident, Expr, Lit};

grammar;

// Exprs

pub Expr: Expr = {
    AppExpr => <>,
    LambdaExpr => <>,
    IfExpr => <>,
    LetExpr => <>,
}

pub AppExpr: Expr = {
    MatchExpr => <>,
    App => <>,
}

App: Expr = {
    <func:AppExpr> <arg:MatchExpr> => Expr::App {func: box func, arg: box arg},
}

pub LambdaExpr: Expr = {
    "\\" <params:Comma1<Ident>> "->" <body:Expr> => Expr::Lambda {params: params, body: box body},
}

pub IfExpr: Expr = {
    "if" <test:Expr> "then" <then_branch:Expr> "else" <else_branch:Expr>
         => Expr::If {test: box test, then_branch: box then_branch, else_branch: box else_branch},
}

pub LetExpr: Expr = {
    "let" <bindings:Comma1<Binding>> "in" <body:Expr>
          => Expr::Let {bindings, body: box body},
}

Binding: (Ident, Expr) = {
    <var:Ident> "=" <val:Expr> => (var, val),
}

MatchExpr: Expr = {
    SimpleExpr => <>,
    // TODO: pattern matching
}

pub SimpleExpr: Expr = {
    VarExpr => <>,
    LitExpr => <>,
    "(" <Expr> ")" => <>,
}

pub VarExpr: Expr = {
    <Ident> => Expr::Var {name: <>.into()},
}

pub LitExpr: Expr = {
    Lit => Expr::Lit {val: <>},
}

pub Lit: Lit = {
    Bool => Lit::Bool(<>),
    Int => Lit::Int(<>),
    Float => Lit::Float(<>),
}

// Terminals

pub Ident: Ident = {
    <s:r"[a-zA-Z][a-zA-Z0-9_]*"> => <>.into(),
}

pub Bool: bool = {
    "true" => true,
    "false" => false,
}

pub Int: i64 = {
    <s:r"[0-9]+"> => i64::from_str(s).unwrap(),
}

pub Float: f64 = {
    <s:r"[0-9]+\.[0-9]+"> => f64::from_str(s).unwrap(),
}

// Helpers

#[inline]
Delim<T, D>: Vec<T> = {
    <v:(<T> <D>)*> <e:T?> => match e {
        None => v.iter().map(|x| x.0.clone()).collect(),
        Some(e) => {
            let mut v = v.iter().map(|x| x.0.clone()).collect::<Vec<_>>();
            v.push(e);
            v
        }
    }
}

#[inline]
Delim1<T, D>: Vec<T> = {
    <v:(<T> <D>)*> <e:T> => {
        let mut v = v.iter().map(|x| x.0.clone()).collect::<Vec<_>>();
        v.push(e);
        v
    }
}

#[inline]
Comma<T> = Delim<T, ",">;

#[inline]
Comma1<T> = Delim1<T, ",">;
