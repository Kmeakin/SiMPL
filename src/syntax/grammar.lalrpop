use std::str::FromStr;
use crate::syntax::ast::*;
use simple_symbol::intern;

grammar;

// Exprs

pub Expr: Expr = {
    LambdaExpr => <>,
    IfExpr => <>,
    LetExpr => <>,
    LetrecExpr => <>,
    CompareExpr => <>,
}

pub MultOp: Expr = {
    <lhs: MultExpr> "*" <rhs: AppExpr>
          => Expr::Binop {lhs: box lhs, rhs: box rhs, op: Binop::IntMul},
    <lhs: MultExpr> "/" <rhs: AppExpr>
          => Expr::Binop {lhs: box lhs, rhs: box rhs, op: Binop::IntDiv},

    <lhs: MultExpr> "*." <rhs: AppExpr>
          => Expr::Binop {lhs: box lhs, rhs: box rhs, op: Binop::FloatMul},
    <lhs: MultExpr> "/." <rhs: AppExpr>
          => Expr::Binop {lhs: box lhs, rhs: box rhs, op: Binop::FloatDiv},
}

pub AddOp: Expr = {
    <lhs: AddExpr> "+" <rhs: MultExpr>
          => Expr::Binop {lhs: box lhs, rhs: box rhs, op: Binop::IntAdd},
    <lhs: AddExpr> "-" <rhs: MultExpr>
          => Expr::Binop {lhs: box lhs, rhs: box rhs, op: Binop::IntSub},

    <lhs: AddExpr> "+." <rhs: MultExpr>
          => Expr::Binop {lhs: box lhs, rhs: box rhs, op: Binop::FloatAdd},
    <lhs: AddExpr> "-." <rhs: MultExpr>
          => Expr::Binop {lhs: box lhs, rhs: box rhs, op: Binop::FloatSub},
}

pub CmpOp: Expr = {
    <lhs: AddExpr> "<" <rhs: AddExpr>
          => Expr::Binop {lhs: box lhs, rhs: box rhs, op: Binop::IntLt},
    <lhs: AddExpr> "<=" <rhs: AddExpr>
          => Expr::Binop {lhs: box lhs, rhs: box rhs, op: Binop::IntLeq},
    <lhs: AddExpr> ">" <rhs: AddExpr>
          => Expr::Binop {lhs: box lhs, rhs: box rhs, op: Binop::IntGt},
    <lhs: AddExpr> ">=" <rhs: AddExpr>
          => Expr::Binop {lhs: box lhs, rhs: box rhs, op: Binop::IntGeq},

    <lhs: AddExpr> "<." <rhs: AddExpr>
          => Expr::Binop {lhs: box lhs, rhs: box rhs, op: Binop::FloatLt},
    <lhs: AddExpr> "<=." <rhs: AddExpr>
          => Expr::Binop {lhs: box lhs, rhs: box rhs, op: Binop::FloatLeq},
    <lhs: AddExpr> ">." <rhs: AddExpr>
          => Expr::Binop {lhs: box lhs, rhs: box rhs, op: Binop::FloatGt},
    <lhs: AddExpr> ">=." <rhs: AddExpr>
          => Expr::Binop {lhs: box lhs, rhs: box rhs, op: Binop::FloatGeq},

    <lhs: AddExpr> "==" <rhs: AddExpr>
          => Expr::Binop {lhs: box lhs, rhs: box rhs, op: Binop::Eq},
    <lhs: AddExpr> "!=" <rhs: AddExpr>
          => Expr::Binop {lhs: box lhs, rhs: box rhs, op: Binop::Neq},
}

pub MultExpr: Expr = {
    AppExpr,
    MultOp,
}

pub AddExpr: Expr = {
    MultExpr,
    AddOp,
}

pub CompareExpr: Expr = {
    AddExpr,
    CmpOp,
}


pub Type: Type = {
    <t1: AtomType> "->" <t2: Type> => Type::Fn(box t1, box t2),
    AtomType => <>,
}

AtomType: Type = {
    "Int" => Type::Int,
    "Bool" => Type::Bool,
    "Float" => Type::Float,
    "(" <Type> ")" => <>,
}

pub AppExpr: Expr = {
    MatchExpr => <>,
    App => <>,
}

App: Expr = {
    <func:AppExpr> <arg:MatchExpr> => Expr::App {func: box func, arg: box arg},
}

pub LambdaExpr: Expr = {
    "\\" <params:Comma1<Param>> "->" <body:Expr> => Expr::Lambda {params, body: box body},
}

Param: Param = {
    <name: Symbol> => Param {name, ann: None},
    <name: Symbol> ":" <ty: AtomType> => Param {name, ann: Some(ty)},
}

pub IfExpr: Expr = {
    "if" <test:Expr> "then" <then:Expr> "else" <els:Expr>
         => Expr::If {test: box test, then: box then, els: box els},
}

pub LetExpr: Expr = {
    "let" <bindings:Comma1<LetBinding>> "in" <body:Expr>
          => Expr::Let {bindings, body: box body},
}

LetBinding: LetBinding = {
    <name:Symbol> "=" <val:Expr> => LetBinding {name, ann: None, val: box val},
    <name:Symbol> ":" <ty: Type> "=" <val:Expr> => LetBinding {name, ann: Some(ty), val: box val},
}


pub LetrecExpr: Expr = {
    "letrec" <bindings:Comma1<LetrecBinding>> "in" <body:Expr>
          => Expr::Letrec {bindings, body: box body},
}

LetrecBinding: LetBinding = {
    <name:Symbol> "=" <val:LambdaExpr> => LetBinding {name, ann: None, val: box val},
    <name:Symbol> ":" <ty: Type> "=" <val:LambdaExpr> => LetBinding {name, ann: Some(ty), val: box val},
}


MatchExpr: Expr = {
    SimpleExpr => <>,
    // TODO: pattern matching
}

pub SimpleExpr: Expr = {
    VarExpr => <>,
    LitExpr => <>,
    "(" <Expr> ")" => <>,
}

pub VarExpr: Expr = {
    <Symbol> => Expr::Var {name: <>.into()},
}

pub LitExpr: Expr = {
    Lit => Expr::Lit {val: <>},
}

pub Lit: Lit = {
    Bool => Lit::Bool(<>),
    Int => Lit::Int(<>),
    Float => Lit::Float(<>),
}

// Terminals

pub Symbol: Symbol = {
    <s:r"[a-zA-Z][a-zA-Z0-9_]*"> => intern(String::from(<>)),
}

pub Bool: bool = {
    "true" => true,
    "false" => false,
}

pub Int: i64 = {
    <s:r"(-|\+)?[0-9]+"> => i64::from_str(s).unwrap(),
}

pub Float: f64 = {
    <s:r"(-|\+)?[0-9]+\.[0-9]+"> => f64::from_str(s).unwrap(),
    <s:r"(-|\+)?[0-9]+\.[0-9]+(e|E)(-|\+)?[0-9]+"> => f64::from_str(s).unwrap(),
}

// Helpers

#[inline]
// List of T delimited by D. Trailing delimiter accepted.
Delim<T, D>: Vec<T> = {
    <v:(<T> <D>)*> <e:T?> => match e {
        None => v.iter().map(|x| x.0.clone()).collect(),
        Some(e) => {
            let mut v = v.iter().map(|x| x.0.clone()).collect::<Vec<_>>();
            v.push(e);
            v
        }
    }
}

#[inline]
// List of atleast 1 T delimited by D. Trailing delimiter accepted.
Delim1<T, D>: Vec<T> = {
    <v:(<T> <D>)+> => {
        let v = v.iter().map(|x| x.0.clone()).collect::<Vec<_>>();
        v
    },
    Delim1NoTrail<T, D> => <>,
}

#[inline]
// List of atleast 1 T delimited by D. Trailing delimiter not accepted.
Delim1NoTrail<T, D>: Vec<T> = {
    <v:(<T> <D>)*> <e:T> => {
        let mut v = v.iter().map(|x| x.0.clone()).collect::<Vec<_>>();
        v.push(e);
        v
    }
}

#[inline]
Comma<T> = Delim<T, ",">;

#[inline]
Comma1<T> = Delim1<T, ",">;

#[inline]
Comma1NoTrail<T> = Delim1NoTrail<T, ",">;
