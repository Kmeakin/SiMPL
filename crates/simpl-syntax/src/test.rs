use crate::grammar;
use insta::assert_debug_snapshot;

#[track_caller]
fn test_parse_ok(src: &str) {
    let parser = grammar::ExprParser::new();
    match parser.parse(src) {
        Ok(ast) => assert_debug_snapshot!(ast),
        Err(e) => {
            eprintln!("{}", e);
            panic!("FAILED TO PARSE")
        }
    }
}

#[test]
fn literal() {
    test_parse_ok("123");
    test_parse_ok("123.456");
    test_parse_ok("true");
    test_parse_ok("false");
}

#[test]
fn var() {
    test_parse_ok("abc");
}

#[test]
fn let_binding() {
    test_parse_ok("let x = 5 in x;");
    test_parse_ok("let x = 5, y = false in x;");
}

#[test]
fn lambda_abstraction() {
    test_parse_ok(r"\() -> 1;");
    test_parse_ok(r"\(x) -> x;");
    test_parse_ok(r"\(x, y) -> y;");
}

#[test]
fn function_application() {
    test_parse_ok(r"f()");
    test_parse_ok(r"f(1, 2)");
    test_parse_ok(r"let id = \(x) -> x; in id(id);")
}
