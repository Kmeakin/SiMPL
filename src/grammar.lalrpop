use std::str::FromStr;
use crate::syntax::ast::{Symbol, Expr, Lit};
use simple_symbol::intern;

grammar;

// Exprs

pub Expr: Expr = {
    AppExpr => <>,
    LambdaExpr => <>,
    IfExpr => <>,
    LetExpr => <>,
    LetrecExpr => <>,
}

pub AppExpr: Expr = {
    MatchExpr => <>,
    App => <>,
}

App: Expr = {
    <func:AppExpr> <arg:MatchExpr> => Expr::App {func: box func, arg: box arg},
}

pub LambdaExpr: Expr = {
    "\\" <params:Comma1<Symbol>> "->" <body:Expr> => Expr::Lambda {params: params, body: box body},
}

pub IfExpr: Expr = {
    "if" <test:Expr> "then" <then:Expr> "else" <els:Expr>
         => Expr::If {test: box test, then: box then, els: box els},
}

pub LetExpr: Expr = {
    "let" <bindings:Comma1<Binding>> "in" <body:Expr>
          => Expr::Let {bindings, body: box body},
}

Binding: (Symbol, Expr) = {
    <var:Symbol> "=" <val:Expr> => (var, val),
}


pub LetrecExpr: Expr = {
    "letrec" <bindings:Comma1<LetrecBinding>> "in" <body:Expr>
          => Expr::Letrec {bindings, body: box body},
}

LetrecBinding: (Symbol, Expr) = {
    <var:Symbol> "=" <val:LambdaExpr> => (var, val),
}


MatchExpr: Expr = {
    SimpleExpr => <>,
    // TODO: pattern matching
}

pub SimpleExpr: Expr = {
    VarExpr => <>,
    LitExpr => <>,
    "(" <Expr> ")" => <>,
}

pub VarExpr: Expr = {
    <Symbol> => Expr::Var {name: <>.into()},
}

pub LitExpr: Expr = {
    Lit => Expr::Lit {val: <>},
}

pub Lit: Lit = {
    Bool => Lit::Bool(<>),
    Int => Lit::Int(<>),
    Float => Lit::Float(<>),
}

// Terminals

pub Symbol: Symbol = {
    <s:r"[a-zA-Z][a-zA-Z0-9_]*"> => intern(String::from(<>)),
}

pub Bool: bool = {
    "true" => true,
    "false" => false,
}

pub Int: i64 = {
    <s:r"[0-9]+"> => i64::from_str(s).unwrap(),
}

pub Float: f64 = {
    <s:r"[0-9]+\.[0-9]+"> => f64::from_str(s).unwrap(),
}

// Helpers

#[inline]
Delim<T, D>: Vec<T> = {
    <v:(<T> <D>)*> <e:T?> => match e {
        None => v.iter().map(|x| x.0.clone()).collect(),
        Some(e) => {
            let mut v = v.iter().map(|x| x.0.clone()).collect::<Vec<_>>();
            v.push(e);
            v
        }
    }
}

#[inline]
Delim1<T, D>: Vec<T> = {
    <v:(<T> <D>)*> <e:T> => {
        let mut v = v.iter().map(|x| x.0.clone()).collect::<Vec<_>>();
        v.push(e);
        v
    }
}

#[inline]
Comma<T> = Delim<T, ",">;

#[inline]
Comma1<T> = Delim1<T, ",">;
