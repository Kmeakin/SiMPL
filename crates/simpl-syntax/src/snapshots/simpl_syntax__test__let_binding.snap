---
source: crates/simpl-syntax/src/test.rs
expression: ast
---
Let(
    LetExpr {
        bindings: [
            Binding {
                var: "x",
                val: Lit(
                    Int(
                        5,
                    ),
                ),
            },
        ],
        body: Var(
            VarExpr(
                "x",
            ),
        ),
    },
)
